src/jsifier.js

```diff
@@ -110,6 +110,20 @@ function JSify(data, functionsOnly) {
         ident: '_' + ident
       });
     });
+    // Exports from the JS runtime may be JS library functions
+    EXPORTED_RUNTIME_METHODS.forEach(function(ident) {
+      // No _ prefix for JS runtime methods, so must be marked with $ in the library.
+      // The first character is ignored by the functionStubs handler, so as a workaround
+      // we can put anything there.
+      ident = '*$' + ident;
+      data.functionStubs.push({
+        intertype: 'functionStub',
+        finalName: ident,
+        ident: ident,
+        ignoreIfMissing: true,
+        exported: true
+      });
+    });
```

Comment:
Is this to provide backwards compatibility, or? If `ccall` goes in the library, does it not already follow the regular route that one uses `-s DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=...` to force it being linked in if not otherwise referenced?

Reply:
Yes, this is partially for backwards compatibility, so that people can keep doing `EXTRA_EXPORTED_RUNTIME_METHODS=['ccall']`. This seems unambiguous, they do want it, and they want it exported. So this seems useful, and not just for backwards compatibility, to avoid needing two redundant flags to both include and export (this allows us to remove a comment in settings.js that "you must both include and export").

Reply:
> So this seems useful, and not just for backwards compatibility, 

As we work towards removing items from preamble/runtime, I don't think we should advertise special mechanisms that we crafted around preamble/runtime to work as general items, but rather, would be good to phase those out as well. So I think retaining support for `EXTRA_EXPORTED_RUNTIME_METHODS=['ccall']` should be considered as backwards compatibility, if we keep it, because `ccall` is not a runtime/preamble function anymore.

> to avoid needing two redundant flags to both include and export (this allows us to remove a comment in settings.js that "you must both include and export").

I think this is the root bug that we should address. Noticed this just now as well, i.e. if I implement a function in a JS library that is not called anywhere, and I want to force it into the build to be exported for my custom outside world JS code to call, I need to add it to both `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE ` and `EXPORTED_FUNCTIONS`, and awkwardly, using both assembly name and C name mangling rules. This occurs in

https://github.com/emscripten-core/emscripten/pull/7667/files#diff-ae40e645f84e893ce5d037b5efc53be6R1239

Ideally we would have only one link+export mechanism that exports both C/C++ compiled functions, and also JS library functions. I.e. if I add in `-s EXPORTED_FUNCTIONS=['_foo']`, it should not matter whether `foo()` is implemented in C/C++ code or in JS library, and it should not matter whether `foo()` had any pre-existing references to it or not in some C/C++/JSlib code - it should always be linked in, and exported.

If above scheme is technically difficult, then second favorite approach would be to have a pair `-s EXPORTED_FUNCTIONS=['_foo']` and `-s EXPORTED_JS_FUNCTIONS=['_foo']`, where first directive links in and exports C/C++ functions, and second directive links in and exports JS library functions (but punt it to the developer to need to know which one it comes from). This is not as good, but probably would do the job as well, it is not hard to reason.

But it never makes sense to have to separately say "link in X" and then "export X". I.e. saying "export X" should always imply saying "link in X", because the export activity should constitute an use of X. CurrentIy adding a JS library function in to EXPORTED_FUNCTIONS means "export X, but give me an error if no C/C++/JSlib code referenced that function so it wasn't linked in via 'natural' means". That does not seem like an useful abstraction.

I wonder to fix this: can we just do in emcc.py frontend something like 

```
DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += EXPORTED_FUNCTIONS
```

(modulo C-vs-asm name mangling and compiler should ignore if some JSlib symbols in DEFAULT_LIBRARY_FUNCS_TO_INCLUDE are not found, (they could have been C/C++ symbols))

or if that's the hard part, i.e. we can't distinguish between nonexisting vs C/C++ linked symbols at that point, then perhaps with `EXPORTED_JS_FUNCTIONS` we could do effectively

```
DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += unmangle(EXPORTED_JS_FUNCTIONS)
EXPORTED_FUNCTIONS += EXPORTED_JS_FUNCTIONS
```

to build a command line directive that does both link + export a symbol?

Reply:
> I wonder to fix this: can we just do in emcc.py frontend something like DEFAULT_LIBRARY_FUNCS_TO_INCLUDE  += EXPORTED_FUNCTIONS

Almost, yeah - that's what this PR does for `EXPORTED_RUNTIME_METHODS`, essentially `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE += EXPORTED_RUNTIME_METHODS`.

The tricky part is to not issue an error if those things don't exist - in your example, an EXPORTED_FUNCTION may have been in compiled code, and in this PR, it may be preamble.